cli/lsp/diagnostics.rs

```diff
@@ -0,0 +1,261 @@
+        {
+          if !references.is_empty() {
+            diagnostics.push((
+              file_id,
+              version,
+              references_to_diagnostics(references),
+            ));
+          } else {
+            diagnostics.push((file_id, version, Vec::new()));
+          }
+        }
+      } else {
+        error!("Missing file contents for: {}", specifier);
+      }
+    }
+  }
+
+  diagnostics
+}
+
+type TsDiagnostics = Vec<diagnostics::Diagnostic>;
+
+fn to_lsp_range(
+  start: &diagnostics::Position,
+  end: &diagnostics::Position,
+) -> lsp_types::Range {
+  lsp_types::Range {
+    start: lsp_types::Position {
+      line: start.line as u32,
+      character: start.character as u32,
+    },
+    end: lsp_types::Position {
+      line: end.line as u32,
+      character: end.character as u32,
+    },
+  }
+}
+
+fn get_diagnostic_message(diagnostic: &diagnostics::Diagnostic) -> String {
+  if let Some(message) = diagnostic.message_text.clone() {
+    message
+  } else if let Some(message_chain) = diagnostic.message_chain.clone() {
+    message_chain.format_message(0)
+  } else {
+    "[missing message]".to_string()
+  }
+}
+
+fn to_lsp_related_information(
+  related_information: &Option<Vec<diagnostics::Diagnostic>>,
+) -> Option<Vec<lsp_types::DiagnosticRelatedInformation>> {
+  if let Some(related) = related_information {
+    Some(
+      related
+        .iter()
+        .filter_map(|ri| {
+          if let (Some(source), Some(start), Some(end)) =
+            (&ri.source, &ri.start, &ri.end)
+          {
+            let uri = Url::parse(&source).unwrap();
+            Some(lsp_types::DiagnosticRelatedInformation {
+              location: lsp_types::Location {
+                uri,
+                range: to_lsp_range(start, end),
+              },
+              message: get_diagnostic_message(&ri),
+            })
+          } else {
+            None
+          }
+        })
+        .collect(),
+    )
+  } else {
+    None
+  }
+}
+
+fn ts_json_to_diagnostics(
+  value: Value,
+) -> Result<Vec<lsp_types::Diagnostic>, AnyError> {
+  let ts_diagnostics: TsDiagnostics = serde_json::from_value(value)?;
+  Ok(
+    ts_diagnostics
+      .iter()
+      .filter_map(|d| {
+        if let (Some(start), Some(end)) = (&d.start, &d.end) {
+          Some(lsp_types::Diagnostic {
+            range: to_lsp_range(start, end),
+            severity: Some(match d.category {
+              diagnostics::DiagnosticCategory::Error => {
+                lsp_types::DiagnosticSeverity::Error
+              }
+              diagnostics::DiagnosticCategory::Warning => {
+                lsp_types::DiagnosticSeverity::Warning
+              }
+              diagnostics::DiagnosticCategory::Suggestion => {
+                lsp_types::DiagnosticSeverity::Hint
+              }
+              diagnostics::DiagnosticCategory::Message => {
+                lsp_types::DiagnosticSeverity::Information
+              }
```

Comment:
Consider implementing `From<diagnostics::DiagnosticCategory> for lst_types::DiagnosticSeverity`